### ISSUE TITLE
 Change default faker locale in factory_boy

### ISSUE THREAD

**Issue by Hacktivista** [FactoryBoy/factory_boy#407]
How can I set the default locale in Python's factory_boy for all of my Factories?

In docs says that one should set it with factory.Faker.override_default_locale but that does nothing to my fakers...

```
import factory
from app.models import Example
from custom_fakers import CustomFakers

# I use custom fakers, this indeed are added
factory.Faker.add_provider(CustomFakers)
# But not default locales
factory.Faker.override_default_locale('es_ES')

class ExampleFactory(factory.django.DjangoModelFactory):
    class Meta:
        model = Example

    name = factory.Faker('first_name')


>>> from example import ExampleFactory
>>> e1 = ExampleFactory()
>>> e1.name
>>> u'Chad'
```

**Comment by rbarrois** [comment]
Indeed: ``factory.Faker.override_default_locale()`` is a decorator and context manager: you can only use it as:

```python
with factory.Faker.override_default_locale('es_ES'):
    e1 = ExampleFactory()
```

Or:

```
@factory.Faker.override_default_locale('es_ES')
def test_foo(self):
    e1 = ExampleFactory()
```

The idea is that changing the default locale is mostly used for a few specific tests while the whole suite doesn't care; hence, having a decorator/context manager ensures that the default locale is restored once those tests have run.

However, this is not very clear in the docs; I think we should improve them.
Alternatively, we could add an extra entrypoint to set the global, initial configuration for Faker (including adding extra providers).

**Comment by elbenfreund** [comment]
I agree with @Hacktivista that an easy (and preferably randomizable) way to change the locale would be a major advantage. Right now I am considering using vanilla faker via a custom `fake` fixture that does just that. It would be nice to have something similar with factory_boy's otherwise lovely faker integration.

If thats something you would be interested in getting a PR please let me know.

**Comment by x-yuri** [comment]
See workaround [here][1].

[1]: https://stackoverflow.com/a/55548725/52499

**Comment by x-yuri** [comment]
I was about to create a new issue. But found this one. Even with some feedback from me :)

I looked into it once again. There are basically 2 ways.

As one might find out, the default locale is taken from [`faker`][1]. But setting the default locale is not covered by the `faker`'s documentation. The good news is that with `faker` it's probably not that important, since usually you specify the locale once, when you instantiate the generator (e.g. `Faker(locale='...')`). I tried looking into it, but first, `faker`'s and `factory-boy`'s default locales are default in different ways. Second, I couldn't easily find a way such that I'd be sure I don't introduce issues.

Another option is to do:

```py
factory.Faker._DEFAULT_LOCALE = '...'
```

after importing `factory`. And apparently before declaring factories. This one will probably work out. But I don't like this underscore in front of the name. And the fact that it's not documented.

@rbarrois Do you think the second way can be considered official? Or do you have any better suggestions?

[1]: https://github.com/FactoryBoy/factory_boy/blob/3.2.0/factory/faker.py#L51

**Comment by rbarrois** [comment]
So, let's look at the different situations where one might want to change the default locale:

- For part of a specific test: this is covered by `with override_default_locale():`
- For a specific test function: this is covered by `@override_default_locale()`
- For a collection of tests grouped in a TestCase subclass: `@override_default_locale()` might not be enough here
- For a whole run of the program / test suite, with a fixed value from run to run
- For a whole run of the program / test suite, with different values between runs

For the last two cases, we'll have to look into the program setup:
- For a program (say, generating a realistic dataset), it is usually easy to add a `with override_default_locale()` within the program's `main()` function
- For a test run, this has to happen within the test runner setup

For Django, I believe that the recommended mode is to:
1. Subclass Django's [`DiscoverRunner`](https://docs.djangoproject.com/en/3.1/topics/testing/advanced/#django.test.runner.DiscoverRunner)
2. Overload its [`run_tests`](https://docs.djangoproject.com/en/3.1/topics/testing/advanced/#django.test.runner.DiscoverRunner.run_tests) function, to wrap the initial method inside a `with override_default_locale()`
3. Point your settings' `TEST_RUNNER` to your class:

```python
# settings.py
TEST_RUNNER = 'myproject.testing.MyTestRunner'

# myproject/testing.py
import factory
from django.conf import settings
from django.util import translation
import django.test.runner

class MyTestRunner(django.test.runner.DiscoverRunner):
    def run_tests(self, test_labels, extra_tests=None, **kwargs):
        with factory.Faker.override_default_locale(translation.to_locale(settings.LANGUAGE_CODE)):
            return super().run_tests(test_labels, extra_tests=extra_tests, **kwargs)
```

**Comment by x-yuri** [comment]
@rbarrois Oh, it appears I've got my case covered. Do you think we can document it? Supposedly by adding a common recipe?

**Comment by robvdl** [comment]
For pytest django I just put it in conftest.py since that loads before any tests start:

```python
import faker.config

faker.config.DEFAULT_LOCALE = "en_NZ"
```

Works a charm, if you check the Faker class, it populates _DEFAULT_LOCALE from there.

**Comment by x-yuri** [comment]
Indeed, just importing `factory` or `faker` doesn't make use of the `faker.config.DEFAULT_LOCALE` value. So sounds safe.

Although I'd like to point out at a slight distinction between `factory-boy`'s and `faker`'s default locales. For `factory-boy` the default locale is a locale used when no locale is specified. That is true for `faker` as well. But in addition `faker` falls back to the default locale if a provider [doesn't support][1] the locale you demanded and doesn't provide its own fallback. For now the only unlocalized provider that doesn't support `en_US` is the `bank` provider, but it provides a [fallback][2].

In other words, changing `factory-boy`'s and `faker`'s default locales might have different effects.

[1]: https://github.com/joke2k/faker/blob/v5.6.1/faker/factory.py#L99-L107
[2]: https://github.com/joke2k/faker/blob/v5.6.1/faker/providers/bank/__init__.py#L9

**Comment by shimizukawa** [comment]
alternative for pytest conftest.py.
```
@pytest.fixture(autouse=True, scope="session")
def set_faker_locale():
    with factory.Faker.override_default_locale("ja_JP"):
        yield
```
